### ISSUE TITLE
Authorization is respecting lazy execution inconsistently

### ISSUE THREAD

**Issue by panthomakos** [rmosolgo/graphql-ruby#1871]
When using the `::authorized?` method on a graphql type, if the result is a lazy evaluation object, it is sometimes evaluated eagerly even when the object itself would not behave the same way. I have created a simple Rails 5.2 app with `graphql (1.8.10)` and `graphql-batch (0.3.10)` to demonstrate this.

I have made the following changes (beyond the `rails new` + addition of gems):

```
# graphql_controller.rb
...
    context = { owner_id: 1 }
```

```
# app/graphql/types/query_type.rb
User = Struct.new(:id)
Permission = Struct.new(:owner_id, :requester_id)

module Loaders
  class UserLoader < ::GraphQL::Batch::Loader
    def perform(ids)
      puts "Loading Users #{ids.inspect}"
      ids.each { |id| fulfill(id, User.new(id)) }
    end
  end
end

module Loaders
  class PermissionLoader < ::GraphQL::Batch::Loader
    def initialize(owner_id)
      @owner_id = owner_id
    end

    def perform(requester_ids)
      puts "Loading #{@owner_id} Permissions #{requester_ids.inspect}"
      requester_ids.each { |id| fulfill(id, Permission.new(@owner_id, id)) }
    end
  end
end

module Types
  class UserType < ::Types::BaseObject
    graphql_name 'User'
    field :id, Int, null: false

    def self.authorized?(object, context)
      ::Loaders::PermissionLoader
        .for(context[:owner_id])
        .load(object.id)
        .then { |permission| true }
    end
  end
end

module Types
  class QueryType < Types::BaseObject
    field :users, [::Types::UserType], null: true do
      argument :ids, [Int], required: true
    end

    def users(ids:)
      ::Loaders::UserLoader.for.load_many(ids)
    end

    field :user, ::Types::UserType, null: true do
      argument :id, Int, required: true
    end

    def user(id:)
      ::Loaders::UserLoader.for.load(id)
    end
  end
end
```

If I run the following query, the STDOUT demonstrates that although the users are all properly lazily evaluated, the permissions objects are not:

```
{
  users(ids: [1,2]) {id}
  first: user(id: 3) { id }
  second: user(id: 4) { id }
}
```

STDOUT:

```
Started POST "/graphql" for 127.0.0.1 at 2018-09-27 00:03:27 -0700
Processing by GraphqlController#execute as */*
  Parameters: {"query"=>"{\n  users(ids: [1,2]) {id}\n  first: user(id: 3) { id }\n  second: user(id: 4) { id }\n}", "variables"=>nil, "operationName"=>nil, "graphql"=>{"query"=>"{\n  users(ids: [1,2]) {id}\n  first: user(id: 3) { id }\n  second: user(id: 4) { id }\n}", "variables"=>nil, "operationName"=>nil}}
Loading Users [1, 2, 3, 4]
Loading 1 Permissions [1, 2, 3]
Loading 1 Permissions [4]
Completed 200 OK in 19ms (Views: 0.2ms | ActiveRecord: 0.0ms)
```

As you can see, all of the user ids are collected into the same `perform`/query while the permissions are eagerly evaluated when checking permissions on `first: user(id: 3)`.

**Comment by rmosolgo** [comment]
Hey, thanks for the detailed issue, sorry I haven't had time to take a close look yet! But it's definitely _supposed_ to work just like lazy fields.

**Comment by panthomakos** [comment]
I would be happy to make an attempt at figuring this out, but would appreciate some guidance regarding where in the code and tests to start looking. I don't know my way around this code base at all.

**Comment by panthomakos** [comment]
Any chance for a fix or some guidance here?

**Comment by rmosolgo** [comment]
Sorry, I don't really have time to help out on this right now :'( I'm giving it all I got to ship the new interpreter at my day job. I took your example and put it in a single file, and reproduced the issue on 1.8.11 and 1.9-dev. 

Here's the script to replicate, I hope I can revisit it soon! 

```ruby 
require "graphql"
require "graphql/batch"
p GraphQL::VERSION

User = Struct.new(:id)
Permission = Struct.new(:owner_id, :requester_id)

module Loaders
  class UserLoader < ::GraphQL::Batch::Loader
    def perform(ids)
      puts "Loading Users #{ids.inspect}"
      ids.each { |id| fulfill(id, User.new(id)) }
    end
  end
end

module Loaders
  class PermissionLoader < ::GraphQL::Batch::Loader
    def initialize(owner_id)
      p "Init for #{owner_id}"
      @owner_id = owner_id
    end

    def perform(requester_ids)
      puts "Loading #{@owner_id} Permissions #{requester_ids.inspect}"
      requester_ids.each { |id| fulfill(id, Permission.new(@owner_id, id)) }
    end
  end
end

module Types
  class UserType < GraphQL::Schema::Object
    graphql_name 'User'
    field :id, Int, null: false

    def self.authorized?(object, context)
      p "Authorized? #{object.id}"
      ::Loaders::PermissionLoader
        .for(context[:owner_id])
        .load(object.id)
        .then { |permission| true }
    end
  end
end

module Types
  class QueryType < GraphQL::Schema::Object
    field :users, [::Types::UserType], null: true do
      argument :ids, [Int], required: true
    end

    def users(ids:)
      ::Loaders::UserLoader.for.load_many(ids)
    end

    field :user, ::Types::UserType, null: true do
      argument :id, Int, required: true
    end

    def user(id:)
      ::Loaders::UserLoader.for.load(id)
    end
  end
end

class Schema < GraphQL::Schema
  query(Types::QueryType)
  use GraphQL::Batch
end

p Schema.execute <<-GRAPHQL, context: { owner_id: 1 }
  {
    users(ids: [1,2]) {id}
    first: user(id: 3) { id }
    second: user(id: 4) { id }
  }
GRAPHQL
```

**Comment by rmosolgo** [comment]
In terms of guidance, the only thing I can think of is to port your failing example above to the graphql test suite, then hunt and peck until a solution is found 😬

**Comment by rmosolgo** [comment]
Ok, this bug is a blocker for my work on the new runtime, so maybe it will get fixed there, and I'll follow up here if I have any clues for backporting the fix! 

I posted a failing test case here: https://github.com/rmosolgo/graphql-ruby/pull/1982

**Comment by rmosolgo** [comment]
Fixed in the interpreter by #1984
